File: 61_rotate_list/rotateRight.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getListLength(self, head):
        cnt = 0
        while head:
            head = head.next
            cnt += 1
        return cnt
        
    def rotateRight(self, head, k):
        length = self.getListLength(head)
        if not head or k == 0: return head
        
        if k > length: k = k % length
        walk_length = length - k
        
        if walk_length == 0 or k == 0: return head
        print walk_length
        
        tmp = head
        while walk_length != 0:
            prev = tmp
            tmp = tmp.next
            walk_length -= 1

        r_head = tmp
        
        while tmp and tmp.next:
            tmp = tmp.next

        tmp.next = head
        prev.next = None

        return r_head


        

File: 61_rotate_list/rotateRight.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
    	ListNode* rotateRight(ListNode* head, int k) {
            ListNode* t1 = head;
            ListNode* t2 = head;
            ListNode* prev = t2;
            int length = getListLength(head);
            if(!length) return NULL; // NULL list
            k %= length; // when k > len(list)
            if(!k) return head; // k == 0: not rotate
            while(--k > 0) t1 = t1 -> next;
            while(t1->next) {
                prev = t2;
                t2 = t2 -> next;
                t1 = t1 -> next;
            }
            prev -> next = NULL;
            t1 -> next = head;
            return t2;
    	}
    private:
        int getListLength(ListNode* head) {
            if(!head) return 0;
            ListNode* tmp = head;
            int res = 0;
            while(tmp) {
                res++;
                tmp = tmp -> next;
            }
            return res;
        }
};

int main() {
    return 0;
}
